### ISSUE TITLE
Thread safety and suggested lifetime patterns of storage classes not documented

### ISSUE THREAD

**Issue by fschmied** [Azure/azure-storage-net#732]
The docs (e.g., https://docs.microsoft.com/en-us/dotnet/api/microsoft.windowsazure.storage.cloudstorageaccount) don't mention anything about the thread safety or suggested lifetime/usage patterns of `CloudStorageAccount`, `CloudBlobClient`, `CloudBlobContainer`, etc.

In a server application, how are those supposed to be created and retained? E.g., `CloudStorageAccount` and `CloudBlobClient` instances as global singletons, `CloudBlobContainer` per call? The sample at https://github.com/Azure-Samples/storage-blobs-dotnet-webapp/blob/master/WebApp-Storage-DotNet/Controllers/HomeController.cs instantiates all these classes per call - is this the suggested usage or is it actually expensive to reinstantiate them every time?

(The "Cloud_Storage_Account_Sample" link at https://docs.microsoft.com/en-us/dotnet/api/microsoft.windowsazure.storage.cloudstorageaccount?view=azure-dotnet doesn't work, BTW, but even if it did, the sample doesn't show how to use those classes in production.)

**Comment by rickle-msft** [comment]
Hi, @fschmied. Thank you for your feedback. I have logged your concern for discussion within our team and we will get back to you with more information as soon as we can.

**Comment by fschmied** [comment]
@rickle-msft Any news on this?

**Comment by rickle-msft** [comment]
We are looking into this now.

**Comment by GavinPG** [comment]
@rickle-msft Has there been any news on this issue? I'd like to hear a definitive response too.

**Comment by moshegutman** [comment]
This had some guidance https://docs.microsoft.com/en-us/azure/azure-functions/manage-connections

## Use static clients

To avoid holding more connections than necessary, reuse client instances rather than creating new ones with each function invocation. .NET clients like the [HttpClient](https://msdn.microsoft.com/library/system.net.http.httpclient(v=vs.110).aspx), [DocumentClient](https://docs.microsoft.com/dotnet/api/microsoft.azure.documents.client.documentclient
), and Azure Storage clients can manage connections if you use a single, static client.

Here are some guidelines to follow when using a service-specific client in an Azure Functions application:

- **DO NOT** create a new client with every function invocation.
- **DO** create a single, static client that can be used by every function invocation.
- **CONSIDER** creating a single, static client in a shared helper class if different functions use the same service.

**Comment by plillevold** [comment]
@moshegutman this is nice guidance from a Azure Functions perspective... but is it true for Azure Storage clients? We don't know. Microsoft needs to step up to the plate and just document this already, specifically for the Azure Storage clients. @rickle-msft ?

**Comment by grik001** [comment]
I would also appreciate some clarification on this.

**Comment by fschmied** [comment]
@rickle-msft Can you confirm the guidance from the Azure Functions docs quoted above, i.e., should we use Azure Storage clients as singletons?

**Comment by meixger** [comment]
@rickle-msft Can we get an update?

**Comment by wboevink** [comment]
+1

**Comment by viktors-telle** [comment]
I am interested in this too.

**Comment by kfarmer-msft** [comment]
While we try to minimize the creation of new HttpClient instances internally (see HttpClientFactory.cs, which implements a single static instance), we cannot guarantee that we won't create new instances in the future based on various factors.

As always, perf questions are best answered by determining what a realistic performance goal for your application is, making a straightforward implementation of it, and then measuring.  I would propose defaulting to a unit-of-work pattern:  create based on logical tasks (eg, "manipulate all these files") rather than lifetime of the application or on single calls.  In my experience (largely informed by database access patterns, where connections aren't cheap), that tends to lead toward a maintainable, reasonably-performant situation.  It will also keep your usage patterns open to cases where you need different clients for different situations.

**Comment by fschmied** [comment]
@kfarmer-msft My question arose from the fact that unit of work scoping (or per call scoping) is not the way an `HttpClient` should be used, as this pattern could cause port exhaustion if many units of work - e.g., web requests - are processed in a short time. So I checked the documentation for the Azure Storage clients and found - nothing.

It's cool that you recommend a unit of work pattern and that apparently this will minimize the creation of `HttpClient` instances at the moment, but I'm uneasy when you say - as I understand it - that you cannot guarantee this will hold for the future. I can't build my application on a "one storage client per unit of work" pattern now and have it run into port exhaustion after updating the Azure client libraries.

> As always, perf questions are best answered by determining what a realistic performance goal for your application is, making a straightforward implementation of it, and then measuring. 

I have no specific performance question. My goal is to use the storage clients in a way that is capable of parallel processing/multi-threading and will not cause port exhaustion. I really think the Azure Storage team needs to document how I can reach this goal. (In the API documentation, not some random Github issue.)

**Comment by grik001** [comment]
@kfarmer-msft  I think @fschmied could not have put it better. Using a Unit of Work pattern in my scenario would also exhaust my application since I am processing thousands of rows per minutes and from your comment it seems you are not suggesting a single static instance.

Would it be possible to get some official documentation on this please?

**Comment by darthkurak** [comment]
I'm also seriously interested in this. For reference, as good example, i could put documentation for mongoDB: https://mongodb.github.io/mongo-csharp-driver/2.7/reference/driver/connecting/ - each "Client" class has re-use information about thread safety. Would be great to have something like this in Account Storage main classes.

**Comment by Peter-Optiway** [comment]
Great they solved it by removing the documentation entirely!
https://docs.microsoft.com/en-us/dotnet/api/microsoft.windowsazure.storage.cloudstorageaccount gives a 404.

**Comment by pentp** [comment]
The namespace was changed in the latest major version (~~Windows~~Azure): https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.storage.cloudstorageaccount?view=azure-dotnet

**Comment by lurumad** [comment]
Any update?

**Comment by ysmoradi** [comment]
It's totally annoying )-:

**Comment by CSurieux** [comment]
+1

**Comment by peledins** [comment]
+1; this would be really great to get this documented.

**Comment by giggio** [comment]
@kfarmer-msft This issue is more than an year old, it would be good to determine the guidance for the lifetime of these services, update the docs, and close the issue.

And I don't like at all the statement:

> we cannot guarantee that we won't create new instances in the future based on various factors

I expect Microsoft to not break compatibility like that between major versions. If you decide to now create a new `HttpClient` for every call, you should do that between major versions, never on the same version, and that should be clearly documented so the users can react accordingly.

**Comment by MHanafy** [comment]
@kfarmer-msft I suggest splitting this issue into two, first is a simple question "are storage client functions thread safe?" this should be easy enough to answer, and for Docs to be updated.

**Comment by paulpce** [comment]
I'm also interested in some documentation around this please.

**Comment by ishepherd** [comment]
Interested.
I have noticed that the total time to fetch blobs seems to be linear rather than parallel, when creating a lot of async fetch tasks at once. Interested in whether a pool of clients is safe and helpful.

**Comment by roshankrishnan7** [comment]
+1

**Comment by ludaval** [comment]
+1

**Comment by paulpce** [comment]
@kfarmer-msft It looks like interest in this issue is increasing, we're up to about one new +1 per day!  Can we please have some action on this? We're designing some mission critical components in a high throughput environment that will be using blob storage - and yes, it will be thoroughly tested -  but we just want some official documentation to make sure we're not setting ourselves up for trouble.

**Comment by felipecruz91** [comment]
+1

**Comment by bojanz27** [comment]
+1

**Comment by larryschlatre** [comment]
+1

**Comment by spshenychnyy-mechanicadvisor** [comment]
+100500

**Comment by PingPongSet** [comment]
+9999999999999999999999999999999999999999999999999999999999999999999

**Comment by PingPongSet** [comment]
If each object is created for each request and not reused, will it cause issue, for example, socket exhaustion (not the same issue) for HttpClient?

**Comment by GavinPG** [comment]
+1729

**Comment by mat-kul** [comment]
Still waiting...

**Comment by zmarty** [comment]
We are internal Microsoft team directly impacted by this question/issue. Is it possible for us to send in our singleton HttpClient to this library, instead of it handling connections on its own?

**Comment by mattwhitfield** [comment]
We could also do with knowing what the lifecycle is here...

**Comment by AGBrown** [comment]
@zmarty can you detail how you are impacted, please? Are you (a) developing a new solution using `CloudStorageAccount`, `CloudBlobClient`, `CloudBlobContainer` etc. and in need of guidance on thread safety and/or lifetime of the created objects; (b) experiencing issues (like exceptions or port exhaustion) potentially related to thread safety and/or lifetime of the created objects; or (c) just looking for a way to inject your own singleton `HttpClient` to pre-empt potential issues?

**Comment by AGBrown** [comment]
Well, a lot of +1, a non-trivial number of watchers and several brief comments asking for help aren't prompting a response so here goes ...

# Nothing in the original issue report has yet been addressed in all this discussion over the last 16 months.

-----------------

### Breaking this issue down:

I agree with the following:

> @kfarmer-msft **I suggest splitting this issue into two, first is a simple question "are storage client functions thread safe?"** this should be easy enough to answer, and for Docs to be updated.
> I'm more interested in this question, as an API consumer I do need to know about thread safety.
> While the other part is still related "lifetime guidance", it might take longer due to the nature of the question.
> I hope this approach can expedite the answer a little bit.

### What documentation do we currently have, if any?

If we look at [Azure / Function / How-to guides / Develop / Manage Connections / Static Clients][manage-connections] it says what @moshegutman [quoted](#issuecomment-417512686) previously [(version-specific link)][manage-connections-1504075]:

> Azure Storage clients can manage connections if you use a single, static client.

but it [also says][manage-connections-1504075-next-steps]: 

> For more information about why we recommend static clients, see [Improper instantiation antipattern][inst].

Which in turn says ([version-specific link][inst-2c51448]):

> examples of broker classes that are relevant to Azure applications ... `QueueClient`

There is no mention of storage classes and whereas `QueueClient` actually has got [documentation on thread-safety and storage patterns][qc-1504075] (if you can stumble across it), storage broker classes don't seem to have corresponding documentation even under the [equivalent documentation section][perf-checklist].

That last page then [goes on to say][inst-2c51448-considerations]:

> - The key element of this antipattern is repeatedly creating and destroying instances of a shareable object. *If a class is not shareable (not thread-safe), then this antipattern does not apply.*
> 
> - The *type of shared resource might dictate whether you should use a singleton or create a pool*. The HttpClient class is designed to be shared rather than pooled. Other objects *might support pooling*, enabling the system to spread the workload across multiple instances.

Finally we have the cited [sample code](https://github.com/Azure-Samples/storage-blobs-dotnet-webapp/blob/d16b7ae/WebApp-Storage-DotNet/Controllers/HomeController.cs) which is just plain bad code and no help at all. (Overwriting a `static` field on every call to `Index`???? :scream: ).

### Why is this documentation not helping?

So here's the kicker: this issue is all about the fact that (other than one [throw-away reference hidden in the azure functions documentation][manage-connections-1504075] with no supporting documentation in the actual documentation for the relevant classes):

1. there is no documentation of thread-safety on the classes such as `CloudBlobClient` or the newer v12 `BlobServiceClient` etc.; and
1. there is no documentation on lifetime or usage patterns for these same classes.

### Other asides in this thread.

To revisit @kfarmer-msft [(well-meaning) comment](#issuecomment-438799569):

> perf questions are best answered by determining what a realistic performance goal for your application is, making a straightforward implementation of it, and then measuring.

I respectfully, but firmly, take the view that this isn't answering the issue. This may answer the OP's "_is it actually expensive to reinstantiate them every time_" question, but not the underlying problem. It may be a performance optimisation to work out if object re-use can improve performance, but the other aspect is that it is fundamental to building a basic application to know how to use objects that own and use expensive network resources in order to address scaling and to prevent resource starvation. <sup>1</sup>

If we are forced to do what you recommend then we are faced with the choice of either (a) never getting a project off the ground as it will become mired in blackbox testing; or (b) risking immediate failures on deployment to production . This is the exact reason why we now have `HttpClient` and `HttpClientFactory` with authoritative guidance on how to use them. There is only one way to use `HttpClient`, not different ways you choose depending on your best guess of usage patterns, performance and expected scale.

### What are our current options, as customers of the storage SDK?

The problem with the current lack of documentation is that we are doomed if we do and doomed if we don't:

- If we "guess" that there will be scaling issues then we choose to use singleton instances, but if they aren't thread-safe then we'll get production time runtime exceptions pretty quickly. 
- If we "guess" that they aren't thread-safe we have to implement appropriate pools (currently by hand, which isn't simple in itself <sup>†</sup>), but if they aren't pool-friendly <sup>‡</sup> then we'll get production time runtime exceptions pretty quickly. 
- If we "guess" they aren't thread-safe or pool-friendly then we have to choose to do one-instance-per-call, but then we might hit port exhaustion or performance issues pretty quickly at production time.

--------------

# **TL;DR**: 

There is no guidance for thread-safety for critical classes in the Azure SDK like `CloudBlobClient` or `CloudBlobContainer` or the newer v12 `BlobServiceClient` etc.. We need a documentation point for each critical type on each of: thread-safety documentation; concurrent async calls; and lifetime recommendations. At the very least we need this guidance for the azure storage classes in a page like [this one][perf-checklist], (although this might be better placed at the type and member level for azure storage SDK classes and it would be nice to have a consistent approach across all the azure SDKs).

---------------

_Footnote_: _When I want to know how to use a type I expect to go to the [type's documentation page like [`BlobContainerClient`] or [`CloudBlobClient`] instead of having to stumble across unrelated pages like [azure functions documentation][manage-connections]. At this point in time most of those type documentation pages offer nothing beyond simple function overload parameter intellisense in the IDE. A thread-safety/concurrency section for types/members as appropriate, and "lifetime and usage pattern" recommendation (singleton, pooled, transient) section for types would completely sort this issue out. It would also handle consistent documentation for legacy versions after major-version, namespace and even assembly name/nuget changes._

<sup>†</sup> _Even `Microsoft.Extensions.ObjectPool` doesn't handle things like `IDisposable` or `IDisposableAsync`, or pooling objects by key, and building async-friendly, disposable friendly, keyed pools isn't a trivial piece of work._

<sup>‡</sup>_Maybe they expire old connections in the background and don't silently stand them up again? So you can get non-deterministic errors at runtime due to stale connections_

<sup>1</sup> _Also it is not the role of a customer of an SDK to blackbox test every class for thread-safety, race conditions, and performance and adjust their usage of the SDK accordingly, and apart from anything else if we were all to do this we'd each be reinventing the wheel every time and it would be a massive waste of global resources)!_

[manage-connections]: https://docs.microsoft.com/en-us/azure/azure-functions/manage-connections#static-clients

[manage-connections-1504075]: https://github.com/MicrosoftDocs/azure-docs/blob/1504075/articles/azure-functions/manage-connections.md#static-clients

[manage-connections-1504075-next-steps]: https://github.com/MicrosoftDocs/azure-docs/blob/1504075/articles/azure-functions/manage-connections.md#next-steps

[inst]: https://docs.microsoft.com/en-gb/azure/architecture/antipatterns/improper-instantiation/#problem-description

[inst-2c51448]: https://github.com/MicrosoftDocs/architecture-center/blob/2c51448/docs/antipatterns/improper-instantiation/index.md#problem-description

[inst-2c51448-considerations]: https://github.com/MicrosoftDocs/architecture-center/blob/2c51448/docs/antipatterns/improper-instantiation/index.md#considerations

[qc-1504075]: https://github.com/MicrosoftDocs/azure-docs/blob/1504075/articles/service-bus-messaging/service-bus-performance-improvements.md#reusing-factories-and-clients

[perf-checklist]: https://github.com/MicrosoftDocs/azure-docs/blob/94e92cf/articles/storage/blobs/storage-performance-checklist.md

[`BlobContainerClient`]: https://docs.microsoft.com/en-gb/dotnet/api/azure.storage.blobs.blobcontainerclient?view=azure-dotnet

[`CloudBlobClient`]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.storage.blob.cloudblobclient?view=azure-dotnet-legacy

**Comment by zmarty** [comment]
> @zmarty can you detail how you are impacted, please? Are you (a) developing a new solution using `CloudStorageAccount`, `CloudBlobClient`, `CloudBlobContainer` etc. and in need of guidance on thread safety and/or lifetime of the created objects; (b) experiencing issues (like exceptions or port exhaustion) potentially related to thread safety and/or lifetime of the created objects; or (c) just looking for a way to inject your own singleton `HttpClient` to pre-empt potential issues?

Couple of goals:
1) Today we use a BlobStorage singleton per target storage account. This seems to be working fine, but I wanted to make sure this is correct. 
2) Furthermore, we have been modifying our code wherever possible to send in a singleton HttpClient to each library that makes external connections. For instance, we also use EventGrid and that library allows us to send in our own HttpClient. Is this possible in the WIndowsAzure.Storage library?

**Comment by ishepherd** [comment]
It appears this is the legacy repo, this issue might get more attention at https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/storage

**Comment by seanmcc-msft** [comment]
Hello everyone, sorry for the very delayed response.

Unfortunately, clients in this repo are not thread safe.  To mitigate this issue, I recommend moving to our new library, located here: https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/storage.  All clients in the new library are immutable and thread safe.

-Sean

**Comment by fschmied** [comment]
Hi @seanmcc-msft,

you still need to *document* the suggested lifetime patterns for, e.g., `BlobServiceClient`, `BlobContainerClient`, and `BlobClient` in the new SDK (v12). Are you supposed to create one instance of these per application, web request, or storage access operation? Will the wrong pattern create resource exhaustion? Or doesn't it matter?

I can't find anything about this on https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/README.md, https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobserviceclient?view=azure-dotnet, https://docs.microsoft.com/en-US/azure/storage/blobs/storage-quickstart-blobs-dotnet, or https://docs.microsoft.com/en-US/azure/storage/blobs/storage-blobs-introduction.

**Comment by fschmied** [comment]
I've created a corresponding issue in the other repo...

https://github.com/Azure/azure-sdk-for-net/issues/8941

**Comment by rabrahamOpti** [comment]
Can this issue, at least, get an assignee?

**Comment by ishepherd** [comment]
@rabrahamOpti The library has moved to a new repo. See https://github.com/Azure/azure-storage-net/issues/732#issuecomment-559277954

**Comment by rabrahamOpti** [comment]
> 
> 
> @rabrahamOpti The library has moved to a new repo. See [#732 (comment)](https://github.com/Azure/azure-storage-net/issues/732#issuecomment-559277954)

Right, but this issue is still open, presumably because the team recognizes the need for documentation here. This should be assigned so it isn't forgotten

**Comment by sahana-1** [comment]
I am trying to access a blob using a singleton CloudBlobClient created with a ManagedIdentity Token: 

var token = await this.activeDirectoryClient.GetManagedIdentityTokenForResource(storageResource);
            TokenCredential tokenCredential = new TokenCredential(token?.AccessToken);
            StorageCredentials storageCredentials = new StorageCredentials(tokenCredential);   
            this.cloudBlobClient = new CloudBlobClient(new Uri(containerEndpoint), storageCredentials);

Getting the token using: this.azureServiceTokenProvider.GetAuthenticationResultAsync(resource).ConfigureAwait(false);

However, after the token expires, using the client I cannot access the blob giving the below expected error: 

_Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature_

The documentation for using managed identity tokens with Azure Storage suggests accessing blob after creation of the client. But if client is singleton, how do we refresh the token. Has anyone experienced this issue?

**Comment by kasobol-msft** [comment]
> Hi @seanmcc-msft Sean McCullough FTE,
> 
> you still need to _document_ the suggested lifetime patterns for, e.g., `BlobServiceClient`, `BlobContainerClient`, and `BlobClient` in the new SDK (v12). Are you supposed to create one instance of these per application, web request, or storage access operation? Will the wrong pattern create resource exhaustion? Or doesn't it matter?
> 
> I can't find anything about this on https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/README.md, https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobserviceclient?view=azure-dotnet, https://docs.microsoft.com/en-US/azure/storage/blobs/storage-quickstart-blobs-dotnet, or https://docs.microsoft.com/en-US/azure/storage/blobs/storage-blobs-introduction.

For V12 SDK please refer to https://devblogs.microsoft.com/azure-sdk/lifetime-management-and-thread-safety-guarantees-of-azure-sdk-net-clients/

**Comment by firephantomassasin** [comment]
@sahana-1  how did you manage to resolve your issue? We are trying to restructure our code to make the CloudBlobClient singleton.

**Comment by kasobol-msft** [comment]
@firephantomassasin This might help https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.storage.auth.tokencredential.-ctor?view=azure-dotnet#Microsoft_Azure_Storage_Auth_TokenCredential__ctor_System_String_Microsoft_Azure_Storage_Auth_RenewTokenFuncAsync_System_Object_System_TimeSpan_ .

**Comment by firephantomassasin** [comment]
@kasobol-msft however we are not using TokenCredential, we are using connection string:

// Retrieve storage account from connection string.
CloudStorageAccount storageAccount = CloudStorageAccount.Parse("connection string");

        // Create the blob client.
        CloudBlobClient blobClient = storageAccount.CreateCloudBlobClient();

**Comment by kasobol-msft** [comment]
@firephantomassasin you should be good then. I was referring to @sahana-1 's issue.
As for making `CloudBlobClient` singleton. One can attempt to do this with some success. However, please be mindful that V11 and prior types like `CloudBlobClient` and `CloudBlob` are stateful. They might have mutable properties and some APIs might mutate the state. Therefore I recommend caution while attempting this.
I strongly recommend to move to V12, where all clients has been designed stateless and can be safely used with any lifespan.

**Comment by wondering639** [comment]
@kasobol-msft 
> I strongly recommend to move to V12, where all clients has been designed stateless and can be safely used with any lifespan.
First of all, thanks for this information. Is this also valid for the `UploadBlobAsync(string blobName, Stream content)` overload? I can't imagine how this can be stateless, doesn't it has to maintain the position of the stream, e.g. what's already read and sent and what's not? Is it safe to use multiple calls `UploadBlobAsync(string blobName, Stream content)`  of one `BlobContainerClient` in parallel? Thanks :-)

**Comment by kasobol-msft** [comment]
@wondering639 It's safe to call `UploadBlobAsync(string blobName, Stream content)` in parallel, each call stack would maintain it's own position. See [here](https://devblogs.microsoft.com/azure-sdk/lifetime-management-and-thread-safety-guarantees-of-azure-sdk-net-clients/)
